libpromises/dbm_migration_lastseen.c

```diff
@@ -182,8 +188,96 @@ static bool LastseenMigrationVersion0(DBHandle *db)
     return !errors;
 }
 
+static bool LastseenMigrationVersion1(DBHandle *db)
+{
+    DBCursor *cursor;
+    if (!NewDBCursor(db, &cursor))
+    {
+        Log(LOG_LEVEL_ERR,
+            "Unable to create database cursor during lastseen migration");
+        return false;
+    }
+
+    char *key;
+    void *value;
+    int key_size, value_size;
+
+    // Iterate through all key-value pairs
+    while (NextDB(cursor, &key, &key_size, &value, &value_size))
+    {
+        if (key_size == 0)
+        {
+            Log(LOG_LEVEL_WARNING,
+                "Found zero-length key during lastseen migration");
+            continue;
+        }
+
+        // Only look for old KeyHostSeen entries
+        if (key[0] != 'q')
+        {
+            // Warn about completely unexpected keys
+            if ((key[0] != 'k') && (key[0] != 'a'))
+            {
+                Log(LOG_LEVEL_WARNING,
+                    "Found unexpected key '%s' during lastseen migration",
+                    key);
+            }
+            continue;
+        }
+
+        KeyHostSeen1 *old_value = value;
+        KeyHostSeen new_value = {
+            .acknowledged = true, // We don't know, assume yes
+            .lastseen = old_value->lastseen,
+            .Q = old_value->Q,
+        };
+
+        if (!DBCursorDeleteEntry(cursor))
+        {
+            Log(LOG_LEVEL_ERR,
+                "Unable to delete version 1 of entry for key '%s' during lastseen migration",
+                key);
+            if (DeleteDBCursor(cursor) == false)
+            {
+                Log(LOG_LEVEL_ERR,
+                    "Unable to close cursor during lastseen migration");
+            }
+            return false;
+        }
+
+        if (!DBCursorWriteEntry(cursor, &new_value, sizeof(new_value)))
+        {
+            Log(LOG_LEVEL_INFO,
+                "Unable to write version 2 of entry for key '%s' during lastseen migration",
+                key);
+            if (DeleteDBCursor(cursor) == false)
+            {
+                Log(LOG_LEVEL_ERR,
+                    "Unable to close cursor during lastseen migration");
+            }
+            return false;
+        }
+    }
+
+    if (!DeleteDBCursor(cursor))
+    {
+        Log(LOG_LEVEL_ERR, "Unable to close cursor during lastseen migration");
+        return false;
+    }
+
+    if (!WriteDB(db, "version", "2", sizeof("2")))
+    {
+        Log(LOG_LEVEL_ERR, "Failed to update version number durling lastseen migration");
+        return false;
+    }
+
+    Log(LOG_LEVEL_INFO, "Migrated lastseen database from version 1 to 2");
```

Comment:
Good question. It would most likely be in a irrecoverable broken state. Not sure how to solve this. Maybe we simply move it to the side and start fresh? What do you guys think?

Reply:
Yup. I don't think there's anything better we could do.

Reply:
Don't we have some rebuild logic that is a bit more than just delete and let rebuild from nothing? Moving to the side was as I recall the last thing we would try.